Odd Issues When Running run.sh

A few of us have seen this issue and we haven't been able to figure what is going on.

@guerler captured the relevant log traces that match what I was seeing exactly. They are this on upload https://gist.github.com/guerler/c29e07ed9bf49e583dc14af19c44c407 possibly stemming from this https://gist.github.com/guerler/6d17b1be30ec5d3aa9c0192a091a342d on startup.

The workaround of skipping run.sh worked for me and @guerler but not for @jdavcs I think:

```
GALAXY_CONFIG_FILE=config/galaxy.yml uvicorn --app-dir lib --factory --port 8080 galaxy.webapps.galaxy.fast_factory:factory
```

I'm using the latest dev branch and have tried with no config, a copy of galaxy.yml.sample, and galaxy.yml.sample with gunicorn preload set to false (from above). I'm on Sequoia 17.5.1 using Python 3.13.2.

Here is the stack trace:

```
7.0.0.1:60814 - "PATCH /api/upload/resumable_upload/e0ef3a61e277457ca7061abed4b968eb HTTP/1.1" 204
[2025-10-28 08:53:47 -0400] [48262] [ERROR] Exception in ASGI application
  + Exception Group Traceback (most recent call last):
  |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/_utils.py", line 79, in collapse_excgroups
  |     yield
  |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/base.py", line 183, in __call__
  |     async with anyio.create_task_group() as task_group:
  |                ~~~~~~~~~~~~~~~~~~~~~~~^^
  |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/anyio/_backends/_asyncio.py", line 781, in __aexit__
  |     raise BaseExceptionGroup(
  |         "unhandled errors in a TaskGroup", self._exceptions
  |     ) from None
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/uvicorn/protocols/http/h11_impl.py", line 403, in run_asgi
    |     result = await app(  # type: ignore[func-returns-value]
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |         self.scope, self.receive, self.send
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |     )
    |     ^
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
    |     return await self.app(scope, receive, send)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/fastapi/applications.py", line 1133, in __call__
    |     await super().__call__(scope, receive, send)
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/applications.py", line 113, in __call__
    |     await self.middleware_stack(scope, receive, send)
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/errors.py", line 186, in __call__
    |     raise exc
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/errors.py", line 164, in __call__
    |     await self.app(scope, receive, _send)
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette_context/middleware/raw_middleware.py", line 94, in __call__
    |     await self.app(scope, receive, send_wrapper)
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/base.py", line 182, in __call__
    |     with recv_stream, send_stream, collapse_excgroups():
    |                                    ~~~~~~~~~~~~~~~~~~^^
    |   File "/Users/greg/.local/share/uv/python/cpython-3.13.2-macos-aarch64-none/lib/python3.13/contextlib.py", line 162, in __exit__
    |     self.gen.throw(value)
    |     ~~~~~~~~~~~~~~^^^^^^^
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/_utils.py", line 85, in collapse_excgroups
    |     raise exc
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/base.py", line 185, in __call__
    |     await response(scope, wrapped_receive, send)
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/base.py", line 229, in __call__
    |     await send({"type": "http.response.body", "body": chunk, "more_body": True})
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette_context/middleware/raw_middleware.py", line 83, in send_wrapper
    |     await send(message)
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/errors.py", line 161, in _send
    |     await send(message)
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/uvicorn/protocols/http/h11_impl.py", line 500, in send
    |     output = self.conn.send(event=h11.Data(data=data))
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/h11/_connection.py", line 538, in send
    |     data_list = self.send_with_data_passthrough(event)
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/h11/_connection.py", line 571, in send_with_data_passthrough
    |     writer(event, data_list.append)
    |     ~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/h11/_writers.py", line 65, in __call__
    |     self.send_data(event.data, write)
    |     ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
    |   File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/h11/_writers.py", line 91, in send_data
    |     raise LocalProtocolError("Too much data for declared Content-Length")
    | h11._util.LocalProtocolError: Too much data for declared Content-Length
    +------------------------------------

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/uvicorn/protocols/http/h11_impl.py", line 403, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        self.scope, self.receive, self.send
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/fastapi/applications.py", line 1133, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/applications.py", line 113, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette_context/middleware/raw_middleware.py", line 94, in __call__
    await self.app(scope, receive, send_wrapper)
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/base.py", line 182, in __call__
    with recv_stream, send_stream, collapse_excgroups():
                                   ~~~~~~~~~~~~~~~~~~^^
  File "/Users/greg/.local/share/uv/python/cpython-3.13.2-macos-aarch64-none/lib/python3.13/contextlib.py", line 162, in __exit__
    self.gen.throw(value)
    ~~~~~~~~~~~~~~^^^^^^^
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/_utils.py", line 85, in collapse_excgroups
    raise exc
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/base.py", line 185, in __call__
    await response(scope, wrapped_receive, send)
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/base.py", line 229, in __call__
    await send({"type": "http.response.body", "body": chunk, "more_body": True})
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette_context/middleware/raw_middleware.py", line 83, in send_wrapper
    await send(message)
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/starlette/middleware/errors.py", line 161, in _send
    await send(message)
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/uvicorn/protocols/http/h11_impl.py", line 500, in send
    output = self.conn.send(event=h11.Data(data=data))
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/h11/_connection.py", line 538, in send
    data_list = self.send_with_data_passthrough(event)
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/h11/_connection.py", line 571, in send_with_data_passthrough
    writer(event, data_list.append)
    ~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/h11/_writers.py", line 65, in __call__
    self.send_data(event.data, write)
    ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
  File "/Users/greg/Work/galaxy/.venv/lib/python3.13/site-packages/h11/_writers.py", line 91, in send_data
    raise LocalProtocolError("Too much data for declared Content-Length")
h11._util.LocalProtocolError: Too much data for declared Content-Length
```

[galaxy.yml](https://github.com/user-attachments/files/23187657/galaxy.yml)

--- a/lib/galaxy/dependencies/pinned-requirements.txt
+++ b/lib/galaxy/dependencies/pinned-requirements.txt
@@ -68,7 +68,7 @@ edam-ontology==1.25.2
 email-validator==2.3.0
 et-xmlfile==2.0.0
 exceptiongroup==1.3.0 ; python_full_version < '3.11'
-fastapi-slim==0.118.3
+fastapi==0.118.3
 filelock==3.19.1 ; python_full_version < '3.10'
 filelock==3.20.0 ; python_full_version >= '3.10'
 fissix==24.4.24

--- a/lib/galaxy/dependencies/pinned-requirements.txt
+++ b/lib/galaxy/dependencies/pinned-requirements.txt
@@ -224,7 +224,7 @@ tomli==2.3.0 ; python_full_version < '3.11'
 tornado==6.5.2
 tqdm==4.67.1
 tuspy==1.1.0
-tuswsgi==0.5.5
+tuspyserver==4.2.0
 typing-extensions==4.15.0
 typing-inspection==0.4.2
 tzdata==2025.2

--- a/lib/galaxy/webapps/galaxy/buildapp.py
+++ b/lib/galaxy/webapps/galaxy/buildapp.py
@@ -11,7 +11,6 @@ from typing import Optional
 from urllib.parse import urljoin
 
 from paste import httpexceptions
-from tuswsgi import TusMiddleware
 
 import galaxy.app
 import galaxy.datatypes.registry
@@ -1084,30 +1083,6 @@ def wrap_in_middleware(app, global_conf, application_stack, **local_conf):
         from galaxy.web.framework.middleware.translogger import TransLogger
 
         app = wrap_if_allowed(app, stack, TransLogger)
-    # TUS upload middleware
-    app = wrap_if_allowed(
-        app,
-        stack,
-        TusMiddleware,
-        kwargs={
-            "upload_path": urljoin(f"{application_stack.config.galaxy_url_prefix}/", "api/upload/resumable_upload"),
-            "tmp_dir": application_stack.config.tus_upload_store or application_stack.config.new_file_path,
-            "max_size": application_stack.config.maximum_upload_file_size,
-        },
-    )
-    # TUS upload middleware for job files....
-    app = wrap_if_allowed(
-        app,
-        stack,
-        TusMiddleware,
-        kwargs={
-            "upload_path": urljoin(f"{application_stack.config.galaxy_url_prefix}/", "api/job_files/resumable_upload"),
-            "tmp_dir": application_stack.config.tus_upload_store_job_files
-            or application_stack.config.tus_upload_store
-            or application_stack.config.new_file_path,
-            "max_size": application_stack.config.maximum_upload_file_size,
-        },
-    )
     # X-Forwarded-Host handling
     app = wrap_if_allowed(app, stack, XForwardedHostMiddleware)
     # Request ID middleware

--- a/lib/galaxy/webapps/galaxy/buildapp.py
+++ b/lib/galaxy/webapps/galaxy/buildapp.py
@@ -8,7 +8,6 @@ import sys
 import threading
 import traceback
 from typing import Optional
-from urllib.parse import urljoin
 
 from paste import httpexceptions
 

--- a/lib/galaxy/webapps/galaxy/fast_app.py
+++ b/lib/galaxy/webapps/galaxy/fast_app.py
@@ -170,9 +170,8 @@ def get_openapi_schema() -> dict[str, Any]:
 def include_tus(app: FastAPI, gx_app):
     config = gx_app.config
     root_path = "" if config.galaxy_url_prefix == "/" else config.galaxy_url_prefix
-    prefix = urljoin(root_path, "api/upload/resumable_upload")
     upload_tus_router = create_tus_router(
-        prefix=prefix,
+        prefix=urljoin(root_path, "api/upload/resumable_upload"),
         files_dir=config.tus_upload_store or config.new_file_path,
         max_size=config.maximum_upload_file_size,
     )

--- a/lib/galaxy/webapps/galaxy/fast_app.py
+++ b/lib/galaxy/webapps/galaxy/fast_app.py
@@ -1,6 +1,7 @@
 from typing import (
     Any,
 )
+from urllib.parse import urljoin
 
 from a2wsgi import WSGIMiddleware
 from fastapi import (
@@ -9,6 +10,7 @@ from fastapi import (
 )
 from fastapi.openapi.constants import REF_TEMPLATE
 from starlette.middleware.cors import CORSMiddleware
+from tuspyserver import create_tus_router
 
 from galaxy.schema.generics import CustomJsonSchema
 from galaxy.version import VERSION
@@ -165,6 +167,24 @@ def get_openapi_schema() -> dict[str, Any]:
     )
 
 
+def include_tus(app: FastAPI, gx_app):
+    config = gx_app.config
+    root_path = "" if config.galaxy_url_prefix == "/" else config.galaxy_url_prefix
+    prefix = urljoin(root_path, "api/upload/resumable_upload")
+    upload_tus_router = create_tus_router(
+        prefix=prefix,
+        files_dir=config.tus_upload_store or config.new_file_path,
+        max_size=config.maximum_upload_file_size,
+    )
+    job_files_tus_router = create_tus_router(
+        prefix=urljoin(root_path, "api/job_files/resumable_upload"),
+        files_dir=config.tus_upload_store_job_files or config.tus_upload_store or config.new_file_path,
+        max_size=config.maximum_upload_file_size,
+    )
+    app.include_router(upload_tus_router)
+    app.include_router(job_files_tus_router)
+
+
 def initialize_fast_app(gx_wsgi_webapp, gx_app):
     root_path = "" if gx_app.config.galaxy_url_prefix == "/" else gx_app.config.galaxy_url_prefix
     app = get_fastapi_instance(root_path=root_path)
@@ -178,6 +198,7 @@ def initialize_fast_app(gx_wsgi_webapp, gx_app):
     include_legacy_openapi(app, gx_app)
     wsgi_handler = WSGIMiddleware(gx_wsgi_webapp)
     gx_app.haltables.append(("WSGI Middleware threadpool", wsgi_handler.executor.shutdown))
+    include_tus(app, gx_app)
     app.mount("/", wsgi_handler)  # type: ignore[arg-type]
     if gx_app.config.galaxy_url_prefix != "/":
         parent_app = FastAPI()

--- a/packages/web_apps/setup.cfg
+++ b/packages/web_apps/setup.cfg
@@ -60,7 +60,7 @@ install_requires =
     SQLAlchemy>=2.0.37,<2.1,!=2.0.41
     starlette
     starlette-context
-    tuswsgi
+    tuspyserver
     typing-extensions
     uvicorn
     uvloop>=0.21.0

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -37,7 +37,7 @@ dependencies = [
     "docutils!=0.17,!=0.17.1",
     "dparse",
     "edam-ontology",
-    "fastapi-slim>=0.111.0,!=0.119.*,!=0.120.0,!=0.120.1,!=0.120.2",  # https://github.com/fastapi/fastapi/pull/13918
+    "fastapi>=0.111.0,!=0.119.*,!=0.120.0,!=0.120.1,!=0.120.2",  # https://github.com/fastapi/fastapi/pull/13918
     "fissix",
     "fs",
     "future>=1.0.0",  # Python 3.12 support
@@ -256,9 +256,6 @@ keep-runtime-typing = true
 constraint-dependencies = [
     "limits>=2.5.0",  # prefer not downgrading this to upgrading packaging
     "scipy>=1.14.1; python_version>='3.10'",  # Python 3.13 support
-]
-override-dependencies = [
-    "fastapi; sys_platform == 'never'",
 ]
 default-groups = []
 extra-index-url = ["https://wheels.galaxyproject.org/simple"]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -101,8 +101,8 @@ dependencies = [
     "starlette-context",
     "svgwrite",
     "tifffile",
-    "tuswsgi",
     "typing-extensions",
+    "tuspyserver",
     "uvicorn!=0.28.0",  # https://github.com/galaxyproject/galaxy/issues/17669
     "uvloop>=0.21.0",  # Python 3.13 support
     "WebOb>=1.8.9",  # Python 3.13 support
@@ -256,6 +256,9 @@ keep-runtime-typing = true
 constraint-dependencies = [
     "limits>=2.5.0",  # prefer not downgrading this to upgrading packaging
     "scipy>=1.14.1; python_version>='3.10'",  # Python 3.13 support
+]
+override-dependencies = [
+    "fastapi; sys_platform == 'never'",
 ]
 default-groups = []
 extra-index-url = ["https://wheels.galaxyproject.org/simple"]

